src/asyncdb_wasm_bindgen_futures.rs

```diff
@@ -0,0 +1,151 @@
+use std::collections::HashMap;
+use std::path::Path;
+
+use async_std::channel::{self, TryRecvError};
+use wasm_bindgen_futures::spawn_local;
+
+use crate::asyncdb::{ReceiverExt, Request, Response, CHANNEL_BUFFER_SIZE};
+use crate::snapshot::Snapshot;
+use crate::{Options, Result, Status, StatusCode, DB};
+
+pub(crate) struct Message {
+    pub(crate) req: Request,
+    pub(crate) resp_channel: channel::Sender<Response>,
+}
+
+/// `AsyncDB` makes it easy to use LevelDB in a async-std runtime.
+/// The methods follow very closely the main API (see `DB` type). Iteration is not yet implemented.
+#[derive(Clone)]
+pub struct AsyncDB {
+    shutdown: channel::Sender<()>,
+    send: channel::Sender<Message>,
+}
+
+impl AsyncDB {
+    /// Create a new or open an existing database.
+    pub fn new<P: AsRef<Path>>(name: P, opts: Options) -> Result<AsyncDB> {
+        let db = DB::open(name, opts)?;
+
+        let (send, recv) = channel::bounded(CHANNEL_BUFFER_SIZE);
+        let (shutdown, shutdown_recv) = channel::bounded(1);
+
+        spawn_local(async move {
+            AsyncDB::run_server_async(db, recv, shutdown_recv, HashMap::new(), 0).await;
+        });
+
+        Ok(AsyncDB { shutdown, send })
+    }
+
+    pub(crate) async fn process_request(&self, req: Request) -> Result<Response> {
+        let (tx, rx) = channel::bounded(1);
+
+        let m = Message {
+            req,
+            resp_channel: tx,
+        };
+        if let Err(e) = self.send.send(m).await {
+            return Err(Status {
+                code: StatusCode::AsyncError,
+                err: e.to_string(),
+            });
+        }
+        let resp = rx.recv().await;
+        match resp {
+            Err(e) => Err(Status {
+                code: StatusCode::AsyncError,
+                err: e.to_string(),
+            }),
+            Ok(r) => Ok(r),
+        }
+    }
+
+    pub(crate) async fn run_server_async(
+        mut db: DB,
+        mut recv: impl ReceiverExt<Message> + Clone + 'static,
+        mut shutdown: impl ReceiverExt<()> + Clone + 'static,
+        mut snapshots: HashMap<usize, Snapshot>,
+        mut snapshot_counter: usize,
+    ) {
+        if let Some(message) = recv.recv().await {
```

Comment:
what's the plan/idea behind this? For `AsyncDB`, it's clear: run the blocking I/O bound code on a background thread to keep the main loop responsive.

Why then implement the background thread using asynchronous I/O? If

1) you don't have real threads in whatever runtime you're running your code, and run this in a background task on Tokio/Async-std/whatever, you can do all DB operations in-line in your tasks, with less overhead (no channel communication etc.);
2) if you do have real threads in your runtime, you can just use the original implementation.

From what I can see here it look like we're dealing with case 1, that's how `spawn_local()` is documented in the `wasm_bindgen_futures` crate.

What am I missing?

Reply:
Yeah, it's a bit weird to be honest (and even weirder when you try an implement an Env that will work in WASM/the browser).

My goal was to get this to work with wasm-bindgen-futures, so that it can run in a WASM environment in the browser. Right now, that would mean in-memory only, although I was starting to poke at what a WASM-based Env would look like to run it e.g. in a browser.

The JS runtime is single-threaded, but uses Promises for asynchronous scheduling, so that the browser doesn't block on waiting for something. Though, working through the Env trait, I think it might be more challenging to implement (e.g. how do we sleep in WASM when we don't have time available), so any wasm32-unknown-unknown target would be in-memory only.

Reply:
(and ultimately it would be backed by IndexedDB, which is LevelDB at it's core so it's a bit of a :spider-man-pointing: situation -- maybe it's better to build what I'm building in IndexedDB directly for WASM and use this for the non-WASM Rust stuff)

Reply:
I agree with your last point: unless the IndexedDB API is so restrictive as to be useless (I don't know it myself), it is probably best to use it directly, and instead abstract LevelDB under a common database interface (instead of abstracting storage below LevelDB). It might save you some time and sweat.